11 cloudformation initial setup

andria-hibe opened the pull request:
Closes #11

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,94 @@
+      ApiId: cf.ref('SlackRouterApi'),
+      Description: 'Integration for Slack Slash Commands',
+      IntegrationType: 'AWS_PROXY',
+      IntegrationUri: cf.join('', [
```
Can you replace this with a `cf.arn`?

andria-hibe replied:
For the IntegrationURI? I followed the [docs about uri](https://docs.aws.amazon.com/apigateway/api-reference/resource/integration/#uri) but couldn't find an easier/cleaner way to do it. This honestly took me awhile to figure out so would appreciate any help.

@dakotabenjamin do you know?

arunasank replied:
Ah, you are right. I can't seem to find a easier way to get the lambda URI. You could combine it with the definition for [`cf.arn`](https://github.com/mapbox/cloudfriend/blob/master/lib/intrinsic.js#L218-L222) if you wish, but this seems to be the right way to do it right now.

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,94 @@
+      FunctionName: 'slack-router-lambda',
+      Handler: 'slack-router-lambda.handler',
+      Role: cf.sub(
+        'arn:aws:iam::${AWS::AccountId}:role/service-role/test-api-lambda-sns-role-6ndj69lt'
```
I think we have to separate out the SNS roles per stack, instead of hard-coding this value, to use CloudFormation in the best possible way. : ) Let's talk about this. cc/ @ramyaragupathy

andria-hibe replied:
Hi @arunasank, what do you mean by separate out the SNS roles per stack? This cfn template does not create any SNS resources.

@ramyaragupathy just to confirm we are creating a new role per Lambda function which I will then use to replace the one that is currently in line 62? The role that's there (`test-api-lambda-sns-role-...`) is the role that you initially created when we were starting the project.

andria-hibe replied:
I've created a separate role for the Lambda function.

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,94 @@
+      Code: {
+        S3Bucket: 'lambda-andria',
+        S3Key: 'slack-router-lambda.zip',
+      },
```
Accept the S3 bucket and key as a parameter, and reference it here.

andria-hibe replied:
I was initially planning on putting the gitsha as a parameter which will then be added to the S3Key to complete it—since I thought it was only the name of the zip file (which will be the gitsha) that will be changing once the automated deployment is setup—like so:
```
const Parameters = {
  LambdaGitSha: {
    Type: 'String',
    Description: 'Git SHA of latest commit for Lambda function',
  },
}
...
Code: {
          S3Bucket: 'stork-us-east-1',
          S3Key: cf.sub('bundles/slack-bots/${LambdaGitSha}.zip'),
      },
```
Do we instead want both the S3Bucket and the full S3Key values to be parameters?

arunasank replied:
Great question. We need to have the bucket and prefix information configurable for both security/flexibility.

andria-hibe replied:
a43f0d8d95895

Added bucket name and gitsha as parameters.

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,94 @@
+        ':',
+        cf.ref('SlackRouterApi'),
+        '/*/*',
+      ]),
```
same as above - `cf.arn` will make your template much cleaner, even though `cf.join` works, technically. : )

andria-hibe replied:
What ARN should we replace this with instead? I struggled with this one too since the docs confused me. It seems like API Gateway has many [different ARN formats](https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html) and looking at that list there wasn't any that matched what I needed so I had to create this long and confusing joint string instead.

Would appreciate your help on this one!

arunasank replied:
This one is a little easier. So, if you look at the code for [`cf.arn`](https://github.com/mapbox/cloudfriend/blob/master/lib/intrinsic.js#L218-L222), the structure they follow for non S3 ARNs is:

`'arn:${AWS::Partition}:${service}:${AWS::Region}:${AWS::AccountId}:${suffix}'`

So, your `${service}` would be `execute-api` and your `${suffix}` would be `cf.join('', [cf.ref('SlackRouterApi'), '/*/*')`. You'd replace `cf.join` with `cf.arn('execute-api', cf.join('', [cf.ref('SlackRouterApi'), '/*/*' ])`

andria-hibe replied:
Done c0ae532e990 Thanks for the help!

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,44 @@
+    accumulator[key] = value
+
+    return accumulator
+  }, {})
```
Nice use of reduce! :raised_hands:

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,44 @@
+  const params = {
+    Message: JSON.stringify(eventBody),
+    Subject: 'SNS from Slack Slash Command',
+    TopicArn: `arn:aws:sns:us-east-1:670261699094:${command}`,
```
Is there a way to pass the account ID and region via environment parameters and remove the hardcoded string here?

dakotabenjamin replied:
is it cf.accountId and cf.region? (https://github.com/mapbox/cloudfriend#pseudo-parameters)

andria-hibe replied:
Done here: https://github.com/hotosm/slack-bots/commit/d6ff1455607912f566f606f5d6148cb9ec818d68

Seems like the AWS Region is actually already an environment variable for Lambdas :)

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,44 @@
+      statusCode: 200,
+      body: 'Something went wrong with your request.',
+      headers: { 'Content-Type': 'application/json' },
+    }
```
In the body, can we also tell them what went wrong? Also, let's return an error code for the statusCode, so that in the lambda logs, it registers as an error, not as a successful run. Can we not return a `Promise.reject` or a more explicit error, instead of returning the error as a successful response?

andria-hibe replied:
80b15c26abeaaa7c5c4c2663294e430e190c7ac2

```
catch (error) {
    console.error(error) // Added to log error to CloudWatch

// this return block is to send back a message to Slack
    return {
      statusCode: 200,
      body:
        ':x: Something went wrong with your request. Please try again and if the error persists, post a message at <#C319P09PB>',
      headers: { 'Content-Type': 'application/json' },
    }
  }
```

I added a `console.error(error)` which will log the error in CloudWatch.

A 200 Status is needed for Slack to post the message. Otherwise Slack merely posts a generic message with no formatting.

See the difference here:
![Screen Shot 2020-07-14 at 19 02 18](https://user-images.githubusercontent.com/54427598/87395068-d1c8fb80-c604-11ea-853c-75a2b233294e.png)
The first error is what the code right now sends back (which is sent using the return block with the status 200) while the second one is what gets sent back if we throw an error (or do a Promise.reject), or send back a 500 status code.

@arunasank @ramyaragupathy let me know what you think we should do. Here is the [Slack docs](https://api.slack.com/interactivity/slash-commands#best_practices) as well and if you scroll up a bit there is a portion regarding **Sending error responses**

arunasank replied:
Oh, sounds great @andria-hibe. I guess, since most of the users are non-developers/community members this definitely makes more sense. :+1:

arunasank commented on package.json:
```diff
@@ -15,5 +15,8 @@
+  },
+  "devDependencies": {
+    "@mapbox/cloudfriend": "^4.2.0"
   }
```
:+1: nice work adding it as a dev-dependency and not a dependency.

arunasank commented on src/health-leaderboard/cloudwatch/health-leaderboard-cfn-template.js:
```diff
@@ -0,0 +1,43 @@
+      Code: {
+        S3Bucket: 'lambda-andria',
+        S3Key: 'health-leaderboard.zip',
+      },
```
Accept the bucket and key as a Parameter.

arunasank commented on src/health-leaderboard/cloudwatch/health-leaderboard-cfn-template.js:
```diff
@@ -0,0 +1,43 @@
+      FunctionName: 'health-leaderboard',
+      Handler: 'health-leaderboard-lambda.handler',
+      Role: cf.sub(
+        'arn:aws:iam::${AWS::AccountId}:role/service-role/test-slack-router-role-4hs4dpro'
```
`cf.arn` again

andria-hibe replied:
@ramyaragupathy 
Thank for flagging.

Once I get confirmation that we'll create a separate role for each Lambda then I'll replace this to reference their own respective role ARN instead. :)

arunasank commented on src/health-leaderboard/health-leaderboard-lambda.js:
```diff
@@ -0,0 +1,91 @@
+  overpassTimestamp,
+  leaderboardTimestamp
+) => {
+  const OSM_EPOCH = 22457216
```
curious - what's the OSM epoch?

andria-hibe replied:
It's the number of minutes from the start of Unix epoch in 1970 until Wed, 12 Sep 2012 06:56:00 UTC. For the latest version of this function I put in a comment to specify this since it IS confusing.

I had to use it for the [augmented diffs](https://wiki.openstreetmap.org/wiki/Overpass_API/Augmented_Diffs#Deriving_augmented_diff_.28.22adiff.22.29_identifier_from_time_interval).

Thanks @ramyaragupathy for teaching me this, there was no way I would have figured it out on my own!

arunasank commented on src/health-tm/health-tm-lambda.js:
```diff
@@ -0,0 +1,82 @@
+const fetch = require('node-fetch')
+
+const TM_STATUS_URL =
+  'https://tasking-manager-tm4-production-api.hotosm.org/api/v2/system/heartbeat/'
```
let's pass these as env variables in the template, so if this changes, the template configuration can be modified, and we won't have to touch the code. : ) Add some sanity checks here for the URL.

arunasank commented on src/health-tm/health-tm-lambda.js:
```diff
@@ -0,0 +1,82 @@
+          text: `:world_map: *Number of Projects Hosted*: ${totalProjects}`,
+        },
+      ],
+    },
```
Nice emoji usage. :raised_hands: :slightly_smiling_face: -- this is the nice part about slack bots!

arunasank commented on src/health-tm/health-tm-lambda.js:
```diff
@@ -0,0 +1,82 @@
+    const slackMessage = {
+      response_type: 'ephemeral',
+      blocks: createBlock(status, mappersOnline, totalProjects, channel_name),
+    }
```
nice - haven't seen ephemeral before. what does it do?

andria-hibe replied:
It makes the message appear only to the user who triggered the command and makes the message disappear if you restart Slack. The opposite is 'in_channel' which will show the message to everyone in the conversation/channel.

@ramyaragupathy might actually be good to discuss if any of them needs to be 'in-channel' since they're all currently 'ephemeral' or if we want to give the users the option to specify if they want it posted for others to see?

arunasank commented on src/health-tm/package.json:
```diff
@@ -0,0 +1,14 @@
+  "license": "ISC",
+  "dependencies": {
+    "node-fetch": "^2.6.0"
+  }
```
is cloudfriend not required in each of the lambda JSONs as well?

andria-hibe replied:
If you have it installed on the root package.json then it already works. :)

arunasank replied:
ah, because it's a dev dependency, and only required for the deployment. Gotcha. :+1:

arunasank commented:
:wave: @andria-hibe GREAT first pass at these functions. Can you also post screenshots of what the slack commands look like on the PR? :-) Would make it really easy to follow the flow of the work.

I have posted some initial comments here. We should definitely write test cases here, both to validate the cloudformation and the initial tests for the funcctions in the lambda scripts. Are you familiar with writing tests for AWS methods? Would you like a walk-through for that?

dakotabenjamin commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,94 @@
+  SlackRouterLambda: {
+    Type: 'AWS::Lambda::Function',
+    Properties: {
+      FunctionName: 'slack-router-lambda',
```
This will need to be unique- add `cf.stackName` with a cf.join.

andria-hibe replied:
@dakotabenjamin I'm confused, do I need to add a tag that's unique? Or the function name has to be unique? We wanted the Lambda functions to be the exact same name as its respective SNS which has the same name as the Slash command (so the `/tm-health` command will have a `tm-health` SNS and Lambda).

I would appreciate some guidance on how to do the tags. I see in the [docs here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html) that CloudFormation automatically creates `aws:cloudformation:stack-name` (among others) as a stack-level tag (which I can see in the Lambda functions although not in the SNS topics) so do I need to create other tags or is it that I put the `aws:cloudformation:stack-name` tag for every resource that accepts it? And I'm not sure where I should put the Name (optional) part?

For the folder name I've actually changed it in each command's branches but forgot to pull the changes to this branch. Thanks for flagging!

arunasank replied:
The `FunctionName` has to be unique. You're absolutely right in your understanding/explanation of the relationship between the lambdas and the respective SNS. However, assume we have a `slack-bots-staging` and `slack-bots-production` stack. Each of those needs to have a different router lambda, which, like Dakota suggested will need to have a "stack-dependent" name to be unique, so the name will be based on whether the stack name is staging/production. You could simply have it be the `${name of the stack}-router-lambda`

As far as creating tags go, you wouldn't have to create them explicitly. AWS has some psuedo-parameters that can be referenced automatically in the CloudFormation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html, so you can simply use a `cf.join` with the `cf.stackName` and a suffix, and that in itself will ensure the name is unique, even if multiple stacks are created.

andria-hibe replied:
cdd3dafc1c3

I tried adding some tags but honestly not sure I did that right. I made a tag specifying what project it's part of, the permissions the Lambda has, and what triggers it.

Can I get some clarification on the tagging policy of HOT? I read up on how to do tagging (like this [AWS paper on tagging best practices](https://d1.awsstatic.com/whitepapers/aws-tagging-best-practices.pdf)) but it's mostly guidelines and seems like it's something different depending on the organisation.

I didn't create a `aws:cloudformation:stackName` tag since the [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html) said it's automatically created and I didn't want to double up on tags but let me know if I need to anyway.

andria-hibe commented:
Thanks @arunasank @dakotabenjamin for your help with this!

You can find photos of the current commands in their respective issues:
- health-tm: #6 
- health-leaderboard: #9  
- osmcha-stats: #15 
- tm-stats: #16 

Aruna I'd appreciate some help on creating tests since I've never tested AWS services. I'm hoping to get through all the changes you both asked for (plus some more tasks I need to finish) this week so maybe next week we can schedule a call?

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,99 @@
+    Type: 'AWS::ApiGatewayV2::Route',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      AuthorizationType: 'NONE',
```
Can you explain why you chose `NONE` over the other options available here? Are the other options not available to implement at the moment?

andria-hibe replied:
The other option is using [JWT Tokens](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-jwt-authorizer.html). Since we're going to be validating all requests that comes in using the Slack secret I wasn't sure if we needed this. I specified `NONE ` though to explicitly state this in case we want to change this.

@ramyaragupathy what do you think? I'm not sure what HOT policy is or what is best practice.

ramyaragupathy replied:
We're good to go with the current implementation with slack secrets - wouldn't want to introduce another layer of complexity

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,50 @@
+    return {
+      statusCode: 200,
+      body: 'Your message is being processed...',
+      headers: { 'Content-Type': 'application/json' },
```
This is really neat btw, especially for commands that take longer to be processed. Nice work! :raised_hands:

andria-hibe commented:
Hi @arunasank @dakotabenjamin 
As per feedback from Aruna, since I've pulled my latest commits to this branch it's now gotten confusing on what is actually needed to be reviewed.

Apologies for that, I was referencing code from different branches and so made the mistake of always pulling my latest commits to whatever branch I'm working on.

For this pull request, you only have to review the files in the root cloudformation folder which contains the `slack-router-lambda.js` and `slack-router-cfn-template.js` files. The individual commands will have their own separate pull requests that I'll make once this one has been merged.

Thanks!

andria-hibe commented:
@arunasank @ramyaragupathy 

Hi both, just realised I had no reason to be so worried about losing files or commits so went ahead and removed the files not up for review in this PR.

I edited both the cfn template (use cloudfriend shortcut to create Lambda function, added permissions to Lambda to access Slack secret in Parameter Store) and the Lambda router (it will now verify all Slack requests before sending the request to its SNS-Lambda pair.

I'm still not 100% on the cfn template so would need more feedback on that. Thanks both!

arunasank commented on .prettierrc:
```diff
@@ -0,0 +1,5 @@
+  "tabWidth": 2,
+  "semi": false,
+  "singleQuote": true
+}
```
any reason you have an `.eslintrc.yml` _and_ a `.prettierrc`? You could add these in the `rules` section of the eslintrc. : )

andria-hibe replied:
Prettier automatically enforces a strict style guide across the code base which ESLint does not. Here's a pretty good [write up](https://prettier.io/docs/en/why-prettier.html) on its benefits.

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,104 @@
+  const ssm = getSSMConnection()
+  const ssmResult = await ssm.getParameter(ssmParams).promise()
+  const slackSecret = ssmResult.Parameter.Value
+
```
You can pass the SSM parameter directly as an environment variable to the lambda using `cf.ref`. Take a look at https://stackoverflow.com/questions/56027070/is-it-possible-to-use-ssm-parameters-in-environment-variables-for-a-lambda for an example.

andria-hibe replied:
You can't assign an SSM Secure String as a Lambda environment variable. You get an error similar to this: `Reason: SSM Secure reference is not supported in: [AWS::Lambda::Function/Properties/Environment/Variables/VAR_NAME]`

Having the Lambda access the SSM parameter that it needs (and only what it needs) also means there will be no risk of CloudFormation logging that secret and follows the principle of least privelege. This way as well, if ever the SSM parameter is updated, there will be no need to update the stack since Lambda will always fetch the latest version of the parameter.

ramyaragupathy replied:
Yeah agree with Andria here to fetch parameter store directly - @arunasank - is there a definite reason we should be accessing the values as env variables?

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,104 @@
+const AWS = require('aws-sdk')
+const crypto = require('crypto')
+
+const region = process.env.AWS_REGION
```
This should also be passed as an env variable.

andria-hibe replied:
Do you want me to pass it as an env variable again?

[AWS_REGION is a reserved environment variable for Lambdas](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) so you can just access it directly without needing to pass it as an env variable.

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,119 @@
+      ),
+    },
+  ],
+  Tags: [
```
Tags are not super essential, unless we are tracking them for keeping track of pricing etc. cc/ @ramyargupathy to take the final call on whether these are required.

ramyaragupathy replied:
@dakotabenjamin - what's your tag suggestion for the lambdas here?

dakotabenjamin replied:
Cloudformation applies some tags automatically, but `Name` would be useful as well.

andria-hibe replied:
e313406 Changed the tags to be Name (name of the command) and Project (slackbot) instead.

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,104 @@
+
+let snsConnection
+function getSNSConnection() {
+  if (snsConnection) return ssmConnection
```
shouldn't this be `return snsConnection`?

andria-hibe replied:
Absolutely right. Thanks for flagging!

arunasank commented on cloudformation/slack-router-lambda.js:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    const eventBody = parseBody(event)
+    const command = decodeURIComponent(eventBody.command).substring(1)
```
Add a comment showing the URL format.

andria-hibe replied:
Sorry, could you clarify? I'm not sure what URL you're referring to.

arunasank commented on cloudformation/slack-router-cfn-template.js:
```diff
@@ -0,0 +1,119 @@
+
+const lambda = new cf.shortcuts.Lambda({
+  LogicalName: 'SlackRouterLambda',
+  Handler: 'cloudformation/slack-router-lambda.handler',
```
As far as I know, the handler _has_ to be named, `index.js` file in the root directory of the zip file. In case you notice errors around "handler not found", try this as the first means to debug.

andria-hibe replied:
This works :) Since the Stork automated code deployment zips the whole repo the handler needs to be accessed through a path since it's not in the root directory. You can assign the handler any name as well (index is the default however).
